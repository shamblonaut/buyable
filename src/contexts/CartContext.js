import { createContext } from "react";

const CartContext = createContext({
  cart: {},
  dispatchCartAction: () => {},
});

export default CartContext;
