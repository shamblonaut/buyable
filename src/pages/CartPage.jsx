import { useEffect, useState } from "react";
import { useOutletContext } from "react-router-dom";
import styled from "styled-components";

import ProductCard from "../components/ProductCard";

const ProductList = styled.ul`
  list-style: none;
  padding: 16px;
  width: 100%;
  display: grid;
  grid-template-columns: repeat(auto-fit, 180px);
  gap: 16px;
  margin-bottom: 64px;
`;

const Heading = styled.h2`
  text-align: center;
  margin-bottom: 16px;
`;

const Info = styled.p`
  text-align: center;
`;

const Price = styled.span`
  font-family: Montserrat, sans-serif;
  font-size: 1.25rem;
  font-weight: 700;
`;

const CartPage = () => {
  const { productsData, cart, setCart } = useOutletContext();

  const [totalPrice, setTotalPrice] = useState(0.0);

  useEffect(() => {
    if (!productsData) return;

    let sum = 0.0;
    productsData.forEach((product) => {
      if (cart[product.id]) {
        sum += product.price * cart[product.id];
      }
    });

    setTotalPrice(sum);
  }, [productsData, cart]);

  return (
    <main>
      <Heading>Your Cart</Heading>

      {productsData && Object.keys(cart).length > 0 ? (
        <>
          <Info>
            Total Price: <Price>${totalPrice}</Price>
          </Info>
          <ProductList>
            {productsData
              .filter((product) =>
                Object.keys(cart).includes(product.id.toString()),
              )
              .map((product) => (
                <li key={product.id}>
                  <ProductCard
                    product={product}
                    cart={cart}
                    setCart={setCart}
                  />
                </li>
              ))}
          </ProductList>
        </>
      ) : (
        <Info>Your cart is empty</Info>
      )}
    </main>
  );
};

export default CartPage;
