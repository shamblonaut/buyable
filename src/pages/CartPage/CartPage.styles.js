import styled from "styled-components";

export const ProductList = styled.ul`
  list-style: none;
  padding: 16px;
  width: 100%;
  display: grid;
  grid-template-columns: repeat(auto-fit, 360px);
  justify-content: center;
  gap: 16px;
  margin-bottom: 32px;
`;
