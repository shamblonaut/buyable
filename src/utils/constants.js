export const AppPosition = Object.freeze({
  HOME: 0,
  SHOP: 1,
  CART: 2,
});
