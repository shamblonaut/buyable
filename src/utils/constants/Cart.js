export const CartAction = Object.freeze({
  ADDED: "added_to_cart",
  REMOVED: "removed_from_cart",
  INCREMENTED: "incremented_cart_quantity",
  DECREMENTED: "decremented_cart_quantity",
  EDITED: "edited_cart_quantity",
  CLEARED: "cleared_cart",
});
