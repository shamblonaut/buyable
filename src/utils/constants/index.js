import { AppPosition } from "./App";

export { AppPosition };
