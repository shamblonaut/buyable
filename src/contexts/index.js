import CartContext from "./CartContext";

export { CartContext };
