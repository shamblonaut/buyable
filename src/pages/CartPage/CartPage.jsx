import { useEffect, useState } from "react";
import { useOutletContext } from "react-router-dom";
import { ShoppingBag } from "lucide-react";

import { AppPosition } from "@/utils/constants";

import { ProductCard } from "@/components";

import { Page, Heading, Info } from "@/styles";

import {
  ProductList,
  InfoTitle,
  CheckoutInfo,
  CheckoutDetails,
  CheckoutLink,
} from "./CartPage.styles";

const CartPage = () => {
  const { productsData, cart, setCart, setAppPosition } = useOutletContext();

  const [subtotal, setSubtotal] = useState(0.0);

  const products = productsData.data;

  const clearCart = () => {
    setCart({});
  };

  useEffect(() => setAppPosition(AppPosition.CART), [setAppPosition]);

  useEffect(() => {
    if (!products) return;

    let sum = 0.0;
    products.forEach((product) => {
      if (cart[product.id]) {
        sum += product.price * cart[product.id];
      }
    });

    setSubtotal(sum);
  }, [products, cart]);

  const tax = subtotal * 0.12;
  const total = subtotal + tax;

  return (
    <Page>
      <Heading>Your Cart</Heading>
      {products && Object.keys(cart).length > 0 ? (
        <>
          <ProductList>
            {products
              .filter((product) =>
                Object.keys(cart).includes(product.id.toString()),
              )
              .map((product) => (
                <li key={product.id}>
                  <ProductCard
                    product={product}
                    cart={cart}
                    setCart={setCart}
                  />
                </li>
              ))}
          </ProductList>
          <CheckoutDetails>
            <InfoTitle>Order Summary</InfoTitle>
            <CheckoutInfo>
              <label>Subtotal: </label>
              <p>${subtotal.toFixed(2)}</p>
            </CheckoutInfo>
            <CheckoutInfo>
              <label>Tax (12%): </label>
              <p>${tax.toFixed(2)}</p>
            </CheckoutInfo>
            <CheckoutInfo>
              <label>Total Price: </label>
              <p>${total.toFixed(2)}</p>
            </CheckoutInfo>
            <CheckoutLink to="/disclaimer" onClick={clearCart}>
              <ShoppingBag /> Place Order
            </CheckoutLink>
          </CheckoutDetails>
        </>
      ) : (
        <Info>Your cart is empty</Info>
      )}
    </Page>
  );
};

export default CartPage;
