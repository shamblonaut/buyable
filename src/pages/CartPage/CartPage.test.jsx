import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen } from "@testing-library/react";

import { CartContext } from "@/contexts";
import { AppPosition } from "@/utils/constants";

import { mockProductList, mockCart } from "@tests/data";
import OutletWrapper from "@tests/components/OutletWrapper";

import CartPage from "./CartPage";

const mockOutletContext = {
  productsData: {
    data: mockProductList,
    error: null,
    loading: false,
  },
  setAppPosition: vi.spyOn(
    {
      setAppPosition: vi.fn(),
    },
    "setAppPosition",
  ),
};

vi.mock("@/components", async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    ProductList: ({ products }) => (
      <ul>
        {products.map((product) => (
          <li data-testid="product" key={product.id}>
            {product.id}
          </li>
        ))}
      </ul>
    ),
  };
});

describe("Cart Page", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders correctly", () => {
    const { container } = render(
      <OutletWrapper context={mockOutletContext}>
        <CartContext.Provider value={{ cart: mockCart, setCart: vi.fn() }}>
          <CartPage />
        </CartContext.Provider>
      </OutletWrapper>,
    );

    expect(container).toMatchSnapshot();
  });

  it("sets the app position correctly on load", () => {
    render(
      <OutletWrapper context={mockOutletContext}>
        <CartContext.Provider value={{ cart: mockCart, setCart: vi.fn() }}>
          <CartPage />
        </CartContext.Provider>
      </OutletWrapper>,
    );

    expect(mockOutletContext.setAppPosition).toHaveBeenCalledWith(
      AppPosition.CART,
    );
  });

  it("shows loading text when data is loading", () => {
    render(
      <OutletWrapper
        context={{
          ...mockOutletContext,
          productsData: {
            data: null,
            error: null,
            loading: true,
          },
        }}
      >
        <CartContext.Provider value={{ cart: mockCart, setCart: vi.fn() }}>
          <CartPage />
        </CartContext.Provider>
      </OutletWrapper>,
    );

    expect(
      screen.getByText("loading products", { exact: false }),
    ).toBeInTheDocument();
  });

  it("shows error text when there is an error in loading the data", () => {
    render(
      <OutletWrapper
        context={{
          ...mockOutletContext,
          productsData: {
            data: null,
            error: new Error("Mock Error"),
            loading: false,
          },
        }}
      >
        <CartContext.Provider value={{ cart: mockCart, setCart: vi.fn() }}>
          <CartPage />
        </CartContext.Provider>
      </OutletWrapper>,
    );

    expect(
      screen.getByText("could not load products", { exact: false }),
    ).toBeInTheDocument();
    expect(
      screen.getByText("mock error", { exact: false }),
    ).toBeInTheDocument();
  });

  it("shows message when cart is empty", () => {
    render(
      <OutletWrapper context={mockOutletContext}>
        <CartContext.Provider value={{ cart: {}, setCart: vi.fn() }}>
          <CartPage />
        </CartContext.Provider>
      </OutletWrapper>,
    );

    expect(
      screen.getByText("cart is empty", { exact: false }),
    ).toBeInTheDocument();
  });

  it("shows only the products present in cart in the product list", () => {
    render(
      <OutletWrapper context={mockOutletContext}>
        <CartContext.Provider value={{ cart: mockCart, setCart: vi.fn() }}>
          <CartPage />
        </CartContext.Provider>
      </OutletWrapper>,
    );

    expect(screen.getAllByTestId("product").length).toBe(
      Object.keys(mockCart).length,
    );
    screen.getAllByTestId("product").forEach((productItem, index) => {
      expect(productItem.textContent).toBe(Object.keys(mockCart)[index]);
    });
  });
});
