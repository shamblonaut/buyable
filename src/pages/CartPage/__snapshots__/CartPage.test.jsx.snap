// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart Page > renders correctly 1`] = `
<div>
  <main
    class="sc-hwkwBN kikxJF"
  >
    <p
      class="sc-kNOymR juelym"
    >
      Your Cart
    </p>
    <div
      class="sc-hdBJTi mLuup"
    >
      <p
        class="sc-iIvHqT hZMqep"
      >
        Order Summary
      </p>
      <span
        class="sc-jxOwhs hqKnvm"
      >
        <label>
          Subtotal: 
        </label>
        <p>
          $
          3743.03
        </p>
      </span>
      <span
        class="sc-jxOwhs hqKnvm"
      >
        <label>
          Tax (12%): 
        </label>
        <p>
          $
          449.16
        </p>
      </span>
      <span
        class="sc-jxOwhs hqKnvm"
      >
        <label>
          Total Price: 
        </label>
        <p>
          $
          4192.19
        </p>
      </span>
      <a
        class="sc-lcItFd inqAEF"
        data-discover="true"
        href="/disclaimer"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-shopping-bag"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 10a4 4 0 0 1-8 0"
          />
          <path
            d="M3.103 6.034h17.794"
          />
          <path
            d="M3.4 5.467a2 2 0 0 0-.4 1.2V20a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V6.667a2 2 0 0 0-.4-1.2l-2-2.667A2 2 0 0 0 17 2H7a2 2 0 0 0-1.6.8z"
          />
        </svg>
         Place Order
      </a>
    </div>
    <ul>
      <li
        data-testid="product"
      >
        1
      </li>
      <li
        data-testid="product"
      >
        3
      </li>
      <li
        data-testid="product"
      >
        5
      </li>
      <li
        data-testid="product"
      >
        7
      </li>
      <li
        data-testid="product"
      >
        9
      </li>
    </ul>
  </main>
</div>
`;
